* Retrieve the instance from deferred.promise
* Methods:
   - then
  - catch
  - finally

* then(successCallback, errorCallback, notifyCallback) – regardless of when the promise was or will be resolved or rejected, then calls one of the success or error callbacks asynchronously as soon as the result is available. The callbacks are called with a single argument: the result or rejection reason. Additionally, the notify callback may be called zero or more times to provide a progress indication, before the promise is resolved or rejected.

* This method returns a new promise which is resolved or rejected via the return value of the successCallback, errorCallback (unless that value is a promise, in which case it is resolved with the value which is resolved in that promise using promise chaining). It also notifies via the return value of the notifyCallback method. The promise cannot be resolved or rejected from the notifyCallback method.

* catch(errorCallback) – shorthand for promise.then(null, errorCallback)

* finally(callback, notifyCallback) – allows you to observe either the fulfillment or rejection of a promise, but to do so without modifying the final value. This is useful to release resources or do some clean-up that needs to be done whether the promise was rejected or resolved. See the full specification for more information.

* New instance is created using $q.defer().promise;

<p data-height="268" data-theme-id="0" data-slug-hash="OyqNXV" data-default-tab="result" data-user="walkingtree" class='codepen'>See the Pen <a href='http://codepen.io/walkingtree/pen/OyqNXV/'>OyqNXV</a> by Walking Tree (<a href='http://codepen.io/walkingtree'>@walkingtree</a>) on <a href='http://codepen.io'>CodePen</a>.</p>
<script async src="//assets.codepen.io/assets/embed/ei.js"></script>

#### Explanation
In the above example when we try to get known location api will get location details like latitude and longitude.

If we enter unknown location api will returns an error and we are handling through errorcallback and showing it to user in nice manner.

If we observe in then there are three methods basically called successCallBack, errorCallBack and notifyCallBack.
Each responds based on response. If success then successCallBack is called and if response is error then errorCallBack is called.
